language/move-stdlib/src/natives/debug.rs

```diff
@@ -3,20 +3,537 @@
+        MoveTypeLayout::Struct(annotated_struct_layout) => Ok(annotated_struct_layout),
+        _ => Err(
+            PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(
+                "Could not convert Type to fully-annotated MoveTypeLayout via NativeContext"
+                    .to_string(),
+            ),
+        ),
+    }
+}
+#[cfg(feature = "testing")]
+fn get_vector_inner_type(ty: &Type) -> PartialVMResult<&Type> {
+    match ty {
+        Type::Vector(ty) => Ok(ty),
+        _ => Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)
+            .with_message("Could not get the inner Type of a vector's Type".to_string())),
+    }
+}
+
+#[cfg(feature = "testing")]
+fn vector_move_value_to_vec_u8(bytes: Vec<MoveValue>) -> PartialVMResult<Vec<u8>> {
+    let mut buf = vec![];
+    for byte in bytes {
+        match byte {
+            MoveValue::U8(u8) => {
+                buf.push(u8);
+            }
+            _ => {
+                return Err(
+                    PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(
+                        "Got not u8 inside bytes field of std::string::String bytes".to_string(),
+                    ),
+                );
+            }
+        }
+    }
+    Ok(buf)
+}
+
+#[cfg(feature = "testing")]
+/// Prints an std::string::String by wrapping it in double quotes and escaping double quotes inside it.
+///
+/// Examples:
+///  - 'Hello' prints as "Hello"
+///  - '"Hello?" What are you saying?' prints as "\"Hello?\" What are you saying?"
+fn print_move_value_as_string(out: &mut String, val: MoveValue) -> PartialVMResult<()> {
+    match val {
+        MoveValue::Vector(bytes) => {
+            let buf = vector_move_value_to_vec_u8(bytes)?;
+
+            let str = String::from_utf8(buf).map_err(|e| {
+                PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(
+                    "Could not parse UTF8 bytes: ".to_string() + e.to_string().as_str(),
+                )
+            })?;
+
+            // We need to escape displayed double quotes " as \" and, as a result, also escape
+            // displayed \ as \\.
+            let str = str.replace('\\', "\\\\").replace('"', "\\\"");
+
+            write!(out, "\"{}\"", str).map_err(HANDLE_FMT_WRITE_ERROR)?;
+        }
+        _ => {
+            return Err(
+                PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(
+                    "Expected std::string::String to have vector<u8> bytes field".to_string(),
+                ),
+            )
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(feature = "testing")]
+fn print_padding_at_depth(out: &mut String, depth: usize) -> PartialVMResult<()> {
+    for _ in 0..depth {
+        // add 2 spaces
+        write!(out, "  ").map_err(HANDLE_FMT_WRITE_ERROR)?;
+    }
+
+    Ok(())
+}
+
+#[cfg(feature = "testing")]
+fn is_vector_u8(vec: &Vec<MoveValue>) -> bool {
+    if vec.is_empty() {
+        false
```

Comment:
An empty vector can't be a `vector<u8>`? Probably just a naming issue, but potentially confusing.

Reply:
Hm, should rename it indeed. The problem is I can't tell if it's a `vector<u8>` because there's no `MoveValue` I can pop to `match` for a `MoveValue::U8` on... Fortunately, I don't need to be able to identify empty `vector<u8>`'s so I just return false.

But this function deserves a clearer name like `is_non_emtpy_vector<u8>`.